helix-core/src/uri.rs

```diff
@@ -0,0 +1,102 @@
+use std::path::{Path, PathBuf};
+
+/// A generic pointer to a file location.
+///
+/// Currently this type only supports paths to local files.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+#[non_exhaustive]
+pub enum Uri {
+    Path(PathBuf),
+}
+
+impl Uri {
+    pub fn from_file_path<P: AsRef<Path>>(path: P) -> Self {
+        Self::Path(path.as_ref().to_path_buf())
+    }
+
+    // This clippy allow mirrors url::Url::from_file_path
+    #[allow(clippy::result_unit_err)]
+    pub fn to_url(&self) -> Result<url::Url, ()> {
+        match self {
+            Uri::Path(path) => url::Url::from_file_path(path),
+        }
+    }
+}
+
+impl From<Uri> for PathBuf {
+    fn from(uri: Uri) -> Self {
+        match uri {
+            Uri::Path(path) => path,
+        }
+    }
+}
+
+impl AsRef<Path> for Uri {
+    fn as_ref(&self) -> &Path {
+        match self {
+            Uri::Path(path) => path,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum UrlConversionError {
+    UnsupportedScheme(url::Url),
+    UnableToConvert(url::Url),
+}
+
+impl std::fmt::Display for UrlConversionError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::UnsupportedScheme(url) => {
+                write!(f, "unsupported scheme in URL: {}", url.scheme())
+            }
+            Self::UnableToConvert(url) => write!(f, "unable to convert URL to file path: {}", url),
+        }
+    }
+}
+
+impl std::error::Error for UrlConversionError {}
+
+impl TryFrom<url::Url> for Uri {
+    type Error = UrlConversionError;
+
+    fn try_from(url: url::Url) -> Result<Self, Self::Error> {
+        if url.scheme() == "file" {
+            url.to_file_path()
+                .map(|path| Self::Path(helix_stdx::path::normalize(path)))
+                .map_err(|_| UrlConversionError::UnsupportedScheme(url))
+        } else {
+            Err(UrlConversionError::UnsupportedScheme(url))
+        }
```

Comment:
this duplication of logic is really unfortunate, but i can see why it's done. some options:
1. leave as is.
2. change the error type to use a `Cow`.
3. bite the cost of cloning the url on the error branch, even if we own it. (these errors are very rare and even when encountered are practically never on the hot path.)
4. eagerly convert the url into its stringified representation since that's all we really want to use it for.
5. change the structure of `UrlConversionError` to something more like the following, so we can have a common url-by-reference method which can delegate the insertion of the url in the error:
```rust
struct UrlConversionError {
    pub source: Url,
    pub reason: UrlConversionErrorReason,
}

enum UrlConversionErrorReason {
    UnsupportedScheme,
    UnableToConvert,
}
```

i think i prefer 3 and 5, but 1 is also reasonable for now.

Reply:
3 is quite reasonable but I'll go with 5 since I don't see us having an error kind where we don't want to include the source URI in the error message. No need to have the Uri on every variant of the enum